### PR TITLE
Polish code in entry.go

### PR DESCRIPTION
**Pull request by wenovus**
- Add a few tests.
- Remove redunant code (`entryCache[n] = e`).
- Remove unreachable code (`if e.Kind == LeafEntry`)
- Improve naming and complexity a little.
- Improve comments

### CODE REVIEW

File: pkg/yang/entry.go [hunk 1]
```diff
@@ -71,9 +71,12 @@ type deviationPresence struct {
 	hasMaxElements bool
 }
 
-// An Entry represents a single node (directory or leaf) created from the
-// AST.  Directory entries have a non-nil Dir entry.  Leaf nodes have a nil
-// Dir entry.  If Errors is not nil then the only other valid field is Node.
+// An Entry represents a single node (directory or leaf) created from the AST
```
Comment (robshakir) [inline pkg/yang/entry.go:74]: Directory, leaf or leaf-list? I wonder whether it's best to say `schema tree node, which can be a directory (containing a subtree) or a leaf node (which contains YANG types that have no children, e.g., leaf, leaf-list)`.

File: pkg/yang/entry.go [hunk 2]
```diff
@@ -1482,25 +1480,18 @@ func (s sortedErrors) Less(i, j int) bool {
 		return false
 	}
 
-	// compare compares field x to see which is less.
```
Comment (robshakir) [inline pkg/yang/entry.go:1485]: Was this unused?

**Review by robshakir — APPROVED** [review]

File: pkg/yang/entry.go (hunk 2 above)
Comment (wenovus) [reply pkg/yang/entry.go:1485]: This is used, I simplified it and took it out of the function.

File: pkg/yang/entry.go (hunk 1 above)
Comment (wenovus) [reply pkg/yang/entry.go:74]: Done.

File: pkg/yang/entry.go [hunk 3]
```diff
@@ -71,9 +71,12 @@ type deviationPresence struct {
...
 
-// An Entry represents a single node (directory or leaf) created from the
-// AST.  Directory entries have a non-nil Dir entry.  Leaf nodes have a nil
-// Dir entry.  If Errors is not nil then the only other valid field is Node.
+// An Entry represents a single node (directory or leaf) created from the AST
+// after applying modifications (i.e. uses, augments, deviations).  There are
+// two high-level types of entries: Directory and Leaf entries. They can be
+// distinguished by whether their "Dir" field is nil.  If Errors is not nil
+// then it means semantic errors existed while converting the AST, in which
+// case the only other valid field is Node.
```
[the remark this answers is not in the document]
Comment (wenovus) [reply pkg/yang/entry.go]: `Errors`. I'm adding it to the comment.

**Comment by wenovus** [comment]
@robshakir for simplicity I put all the PRs in a chain, so can't merge until the previous ones have been approved. Starts at #160.

**Pull request merged**